File: src/supernode/src/main/java/com/dragonflyoss/dragonfly/supernode/service/preheat/Preheater.java
package com.dragonflyoss.dragonfly.supernode.service.preheat;

import com.dragonflyoss.dragonfly.supernode.common.domain.PreheatTask;

/**
 * @author lowzj
 */
public interface Preheater {
    /**
     * The type of this preheater
     *
     * @return String
     */
    String type();

    /**
     * Create a worker to preheat the task.
     *
     * @param task preheat task information
     * @param service preheat service
     * @return BaseWorker
     */
    BaseWorker newWorker(PreheatTask task, PreheatService service);

    /**
     * cancel the running task
     *
     * @param id the preheat task id
     */
    void cancel(String id);

    /**
     * remove a running preheat task
     *
     * @param id the preheat task id
     */
    void remove(String id);
}

File: supernode/daemon/mgr/preheat/preheater.go
/*
 * Copyright The Dragonfly Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package preheat

import (
	"github.com/dragonflyoss/Dragonfly/supernode/daemon/mgr"
)

type Preheater interface {
	/**
	 * The type of this preheater
	 */
	Type() string

	/**
	 * Create a worker to preheat the task.
	 */
	NewWorker(task *mgr.PreheatTask , service *PreheatService ) IWorker

	/**
	 * cancel the running task
	 */
	Cancel(id string)

	/**
	 * remove a running preheat task
	 */
	Remove(id string)
}

func GetPreheater(typ string) Preheater {
	return preheaterMap[typ]
}

func RegisterPreheater(typ string, preheater Preheater) {
	preheaterMap[typ] = preheater
}

var preheaterMap = make(map[string]Preheater)